Disable keybindings when GUI has focus

psykzz opened the pull request:
Fixes #104 
Currently has a small bug, where if you close the MenuWindow with the [x] in the top right, focus is not restored to keybindings.

You can press ESC to clear it and keep moving, but this isn't ideal.

Video for demonstration - https://www.youtube.com/watch?v=bMHi0X2gvQ0

BobdaBiscuit commented:
I don't think disabling movement with menus open is a good gameplay decision from all standpoints.

Is there any way to make this only happen when a text box is in focus?

psykzz commented:
Updated to now block chat, and also fixed bug of keybindings not enabling again.
Note: Focus can be set on disposing elements, call RemoveFocus after Dispose

ComicIronic commented:
I have some things to say about how this is implemented, but my main issue is with the type-checking. It's totally inappropriate in an OO language.

ComicIronic commented on SS14.Client.Services/State/States/GameScreen.cs:
```diff
@@ -373,6 +374,9 @@ private void InitializeGUI()
             UserInterfaceManager.AddComponent(_menuButton);
+
+            // Ensure nothing has focus
+            UserInterfaceManager.RemoveFocus();
```
Why would anything have focus at this point anyways? Does the UserInterfaceManager give new components focus?

Silvertorch5 replied:
Everything in usterinterfacemanager starts with no focus so its unnecessary in my opinion

psykzz replied:
It was more a precaution and a way to ensure state.

ComicIronic replied:
It adds more side-effects to the method and mixes up responsibility. I feel like this could cause confusion in the future.

psykzz commented:
@ComicIronic how would you do the type-checking? Do you have an example?

ComicIronic commented:
Do it polymorphically - either extend the current interface or make a new one, and give it `EnableKeyBindings { get; }`, and set it based on that.

psykzz commented:
@ComicIronic something like this?

ComicIronic commented on SS14.Client.Services/UserInterface/UserInterfaceManager.cs:
```diff
@@ -378,6 +384,34 @@ public virtual bool KeyDown(KeyEventArgs e)
+                if (_currentFocus != null)
+                {
+                    // Handle special case where we are mid-typing.
+                    if (_currentFocus.GetType() == typeof(Chatbox) 
```
You still have this, which is what I have a problem with. The UI manager shouldn't have to care about special rules for each part of the UI.

psykzz commented:
Im going to work on completing the rest of this PR then later tonight.

psykzz commented:
This should be ready for merge now

psykzz commented:
Conflicts resolved for now
